compiler/rustc_lint/src/reexports.rs

```diff
@@ -0,0 +1,82 @@
+use crate::lints::UselessAnonymousReexportDiag;
+use crate::{LateContext, LateLintPass, LintContext};
+use rustc_hir::def::DefKind;
+use rustc_hir::def_id::DefId;
+use rustc_hir::{Item, ItemKind, UseKind};
+use rustc_middle::ty::Visibility;
+use rustc_span::symbol::kw;
+use rustc_span::Span;
+
+declare_lint! {
+    /// The `useless_anonymous_reexport` lint checks if anonymous re-exports
+    /// are re-exports of traits.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(useless_anonymous_reexport)]
+    ///
+    /// mod sub {
+    ///     pub struct Bar;
+    /// }
+    ///
+    /// pub use self::sub::Bar as _;
+    /// # fn main() {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Anonymous re-exports are only useful if it's a re-export of a trait
+    /// in case you want to give access to it. If you re-export any other kind,
+    /// you won't be able to use it since its name won't be accessible.
+    pub USELESS_ANONYMOUS_REEXPORT,
+    Warn,
+    "useless anonymous re-export"
+}
+
+declare_lint_pass!(UselessAnonymousReexport => [USELESS_ANONYMOUS_REEXPORT]);
+
+fn emit_err(cx: &LateContext<'_>, span: Span, def_id: DefId) {
+    let article = cx.tcx.def_descr_article(def_id);
+    let desc = cx.tcx.def_descr(def_id);
+    cx.emit_spanned_lint(
+        USELESS_ANONYMOUS_REEXPORT,
+        span,
+        UselessAnonymousReexportDiag { article, desc },
+    );
+}
+
+impl<'tcx> LateLintPass<'tcx> for UselessAnonymousReexport {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        if let ItemKind::Use(path, kind) = item.kind &&
+            !matches!(kind, UseKind::Glob) &&
+            item.ident.name == kw::Underscore &&
+            // We only want re-exports. If it's just a `use X;`, then we ignore it.
```

Comment:
Why is this important?
A private `use some::Struct as _` is equally useless.

The original suggestion from @cjgillot to use `UnusedImportCheckVisitor` and report this early looks like the right direction.

This case doesn't need a separate lint, existing `unused_imports` should be perfectly fine.

Reply:
This lint multiplies entities without necessity, I think it is better to remove it before it hit stable.

Reply:
(I noticed this while auditing uses of imports in HIR for https://github.com/rust-lang/rust/pull/109176#issuecomment-1477731569.)

Reply:
> This lint multiplies entities without necessity

How so?

> I think it is better to remove it before it hit stable.

Replaced, sure. Otherwise please don't remove it altogether.

Reply:
>How so?

Is there any scenario when you'd want to enable/deny reporting this case specifically, separately from other unused imports? I don't think so.

>Replaced, sure. Otherwise please don't remove it altogether.

Yeah, I mean merge into `unused_imports`, not remove outright.

Reply:
(Also, the lint name doesn't match [conventions](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints).)